implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,611 @@
+    Message.with_local_metadata(msg, Map.merge(additional, %{identity: i, identity_id: id}))
   end
 
-  defp setup_callback_route(options, data) do
-    case Keyword.get(options, :callback_route) do
-      nil ->
-        {:ok, data}
+  defp handle_outer_message_impl(message, %Channel{channel_state: %Established{} = e} = state) do
+    message = Message.forward(message)
+    message = Message.set_onward_route(message, state.peer_identity_route ++ message.onward_route)
 
-      callback_route when is_list(callback_route) ->
-        {:ok, Map.put(data, :callback_route, callback_route)}
+    with {:ok, encrypt_st} <-
+           send_over_encrypted_channel(
+             message,
+             e.encrypt_st,
+             state.peer_route,
+             state.inner_address
+           ) do
+      {:ok, %Channel{state | channel_state: %Established{e | encrypt_st: encrypt_st}}}
     end
   end
 
-  defp setup_extra_init_payload(options, data) do
-    case Keyword.get(options, :extra_init_payload) do
-      nil -> {:ok, data}
-      extra_init_payload -> {:ok, Map.put(data, :extra_init_payload, extra_init_payload)}
+  defp send_over_encrypted_channel(message, encrypt_st, peer_route, inner_address) do
+    with {:ok, encoded} <- Wire.encode(message),
+         {:ok, ciphertext, encrypt_st} <- Encryptor.encrypt("", encoded, encrypt_st) do
+      ciphertext = :bare.encode(ciphertext, :data)
+      envelope = %{onward_route: peer_route, return_route: [inner_address], payload: ciphertext}
+      Router.route(envelope)
+      {:ok, encrypt_st}
     end
   end
 
-  defp setup_authorization(options, data) do
-    case Keyword.get(options, :authorization, []) do
-      authorization when is_list(authorization) ->
-        plaintext_address = Map.fetch!(data, :plaintext_address)
-        ciphertext_address = Map.fetch!(data, :ciphertext_address)
-        data = Map.put(data, :all_addresses, [plaintext_address, ciphertext_address])
-        {:ok, Ockam.Worker.update_authorization_state(data, plaintext_address, authorization)}
-
-      other ->
-        {:error, {:authorization_not_list, other}}
+  defp check_trust(policies, identity, contact, contact_id) do
+    with {:ok, identity_id} <- Identity.validate_identity_change_history(identity) do
+      TrustPolicy.from_config(policies, %{id: identity_id, identity: identity}, %{
+        id: contact_id,
+        identity: contact
+      })
     end
   end
 
-  # sets a key establishment protocol and calls its setup
-  defp setup_key_establishment_protocol(options, data) do
-    case Keyword.get(options, :key_establishment_protocol, XXKeyEstablishmentProtocol) do
-      XXKeyEstablishmentProtocol ->
-        data = Map.put(data, :key_establishment_protocol, XXKeyEstablishmentProtocol)
-        data.key_establishment_protocol.setup(options, data)
+  defp spawner_options(opts) do
+    worker_opts =
+      opts
+      |> Keyword.drop([:address, :authorization])
+      |> Enum.map(fn
```

Comment:
Nit: Almost zero practical difference in the outcome but slightly more intention-revealing IMO. The major distinction is it de-duplicates on purpose, which I'm not sure is benign/good/bad for us here.
```suggestion
      |> Keyword.new(fn
```